### ISSUE TITLE
Proposal for some organisational changes

### ISSUE THREAD

**Issue by TimonPost** [r3bl-org/r3bl_rs_utils#5]
# Usecase

As a developer, I was trying to understand the project and found myself wondering and confused by how the `utils` and `cmdr` repo were named and what their responsibilities are. After cloning the project and taking a closer look it seemed like `r3bl_utils` contained the core code of this awesome library and the `cmdr` repo contained examples.

# Confusing Aspect

- Utils to me means a utility, or something extra, its a bunch of helper functions/structs. Thus by that token, I was not expecting it to contain the core logic of this project and hence why the confusion arose. 
- Making some configuration changes/code changes in one project doesn't affect the other projects. It invokes duplication in maintaining related projects in separate repos. 

# Proposal

Something like:
```
r3bl-org (organisation)
    - r3bl (workspace)
             -  r3bl-core (crate)
             -  r3bl-macro (crate)
             -  r3bl-cmdr (crate)
```

## Single Workspace

I have seen many projects keeping the related code bases inside a single workspace. It is very common in the rust ecosystem to keep code related to a single project close to each other and within the same workspace. This provides many benefits such as:
- Single CI 
- Single project configuration (rustfmt, clippy, vscode, etc.)
- Single issue, PR boards. 
- Single landing page for users to land on.
- Easier to start coding and testing (at the moment the `cmdr` project refers to a dependency to `utils` by local path but one needs to assure it exists first)
- Workspace-wide configurations

## How

The following steps would have to be taken

- Rename this repo from `r3bl_rs_utils` to `r3bl`, this is the project name and hence a natural fit for this repo. This project already has 22 stars so would be nice to reuse this repo and just rename. 
- Move a clone of  `r3bl_rs_utils` to a new repository
   - Update the README of this new repo with bold letters **DEPRECATED** 
   - Release a new cargo version of this  `r3bl_rs_utils` crate that shows it is deprecated.
- Move `cmdr` into the `r3bl` repo and create a workspace
- Depreciate the `cmdr` repository (or delete it as it likely doesnt have many users yet)
- Update the `Cargo.toml` and rename the main project to `r3bl`
- Release `r3bl`.

## Discussion

You would end up having to deprecate `r3bl_rs_utils` but since it isn't used that much yet it should not be a problem. Those kinds of changes is better early than too late.  Just publish a new version with a README saying it's deprecated and link to the new project, this repo, under its new name. 

What do you think?

**Comment by nazmulidris** [comment]
Thanks for the feedback. Unfortunately I didn't get a chance to explain earlier what `r3bl-cmdr` is and what relationship is to `r3bl_rs_utils`.

In the meantime, I've added some more context around what the relationship between these 2 projects are here to help folks get started - https://github.com/r3bl-org/r3bl_rs_utils/blob/main/CONTRIBUTING.md#how-are-r3bl-repos--crates-organized

Since the project is at a very early stage right now, the following parts are going to evolve.
- `r3b_rs_utils` was always meant to be a library. This is currently ready for us.
    - It was meant to be separate from `r3bl-cmdr`. 
    - The idea is that other projects can use the `redux` mod, `tui` mod, `macro` crate & `core` crate w/out having any involvement w/ `r3bl-cmdr`.
- `r3bl-cmdr` is meant to be a user facing application *which is currently totally undeveloped*. This is not released yet.
    - It is a placeholder at the moment, and will be replaced over the coming wks/months w/ code that will end up becoming the 1st version of the product. 
    - Releases will eventually be cut from this & binaries released via `brew` (including `cargo`) or other software distro mechanisms.

You brought up another point about the name of the library itself. A better name, thinking about it now, would have been `r3bl-core`. However, the underlying functionality would be the same. It would still have been a library, that isn't meant to be a heavy framework. Rather a set of loosely coupled, strongly coherent, modules that can be assembled & work together as needed. So you can a la carte pick `tui` module, but not the `redux` module, or vice versa.

> The reason the name is `r3bl_rs_utils` is because a long time ago (~2yrs ago) I was building a NodeJS & TS version of it, since I was thinking about using NodeJS, TS, Ink3, React. Here's a link to that one - https://github.com/r3bl-org/r3bl-ts-utils. About 6 mos ago, I learned Rust & essentially rewrote all of it in Rust (but w/ a layout, styling, redux engine from scratch).

**Comment by TimonPost** [comment]
Thanks for the elaboration on the historical reasons and meaning of those libraries! Asside from the provided argumentation, what do you think of the proposal.

```
- r3bl-rs
       - src/ or r3bl (contains main access code to your library)
       - r3bl-tui
       - r3bl-redux
       - r3bl-core 
       - r3bl-macro
       - r3bl-cmdr (how users access the library as a binary)
       - r3bl-utils (collection of utilities) (might be merged into r3bl code base. 
       - etc..
```

The important thing here is that they all exist within a single workspace with the benefits as described above.  This is a common way to organize related rust projects. Tokio, rand, syn, quinn, etc... to name a few.

**Comment by nazmulidris** [comment]
_tl;dr:_ Thank you for the refinement of your proposal & your passion to improve the organization of things 🙏🏽. `Tokio`, `rand`, `syn` etc. are all libraries. `r3bl-cmdr` is not; it is a user facing product (a suite of apps in a *binary crate* that may be distributed via brew, apt, crates.io, whatever). By design, it is meant to be separate from the underlying library. `r3bl_rs_utils` is a *library crate* & I agree w/ you that more extensive use of workspaces are a great idea to modularize it further (it already uses workspaces today). While this library is created in the service of `r3bl-cmdr`, it is designed to be used by _any other project_ as well (and not as an afterthought). They are separate by design. You can look at `nushell` (akin to `r3bl-cmdr`) & `reedline` (akin to `r3bl_rs_utils`) as an example of this separation pattern.

I agree w/ you that workspaces are a good thing. Currently, there is a workspace file [here](https://github.com/r3bl-org/r3bl_rs_utils/blob/main/Cargo.toml#L67) that binds the `macro` and `core` crates together into a single workspace. These 3 crates are part of the same workspace: `r3bl_rs_utils`, `r3bl_rs_utils_core`, `r3bl_rs_utils_macro`.

However, putting `r3bl-cmdr` under the `r3bl_rs_utils` namespace is "inverted" from the perspective of our product design & goals. And thus the proposal to do so is a nonstarter.

Having said that, I do like the part of your proposal that contains the idea of even more separation inside the `r3bl_rs_utils` library crate itself. Here's a riff off your proposal:

```text
r3bl_rs_utils     (crate & workspace)
  - src/ 
  - core          (multiple lib crate)
  - macro         (multiple lib crate)
  - tui           (multiple lib crate)
  - redux         (multiple lib crate)
  - etc..

r3bl-cmdr         (crate)
  - src/ ...      (one of the deps is r3bl_rs_utils)
```

cc: @shayes12111

**Comment by TimonPost** [comment]
Thanks for the extended elaboration, sorry I looked at the `cargo.toml` but had not noted the 'workspace' section whoops. I am not very well known for the insides and outsides of the whole architecture of the project so forgive me for my ignorance. It appeared to me from the current code that cmdr was more or less a demo app to implement certain examples of the utils library. But with your elaboration, it makes it more clear that this is going to be the binary through which users will build TUI apps. And I can see the reasons for separating it from the repo and workspace. The separation between framework and app(s) does make sense. 

Its epic that it already uses workspaces! I like your riff off of your proposal of my proposal. Think it makes it more modularized and well separated.

**Comment by nazmulidris** [comment]
Thank you for pointing me to the workspace resource [here](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html#creating-the-second-package-in-the-workspace). I will definitely lean into it further & make the internals of `r3bl_rs_utils` use many more modules & workspaces to reduce build times.

It is confusing right now, since r3bl-cmdr is empty. So it makes perfect sense where you were coming from 😄 .

Thank you again for your passion & commitment to making this thing more organized not just to reduce cognitive load for someone trying to grok this for the first time, but also keeping an eye on reducing build times.

Once r3bl-cmdr has more stuff in it, we will be sure to have lots of documentation in there to explain this separation so that it is apparent. Whether someone jumps into cmdr first or the library first, we will put an emphasis in the docs to explain the context.
